metrics-core/src/lib.rs

```diff
@@ -32,24 +32,195 @@
+    pub fn into_parts(self) -> (ScopedString, ScopedString) {
+        (self.0, self.1)
+    }
+}
+
+/// A metric key.
+///
+/// A key always includes a name, but can optional include multiple labels used to further describe
+/// the metric.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Key {
+    name: ScopedString,
+    labels: Vec<Label>,
+}
+
+impl Key {
+    /// Creates a `Key` from a name.
+    pub fn from_name<N>(name: N) -> Self
+    where
+        N: Into<ScopedString>,
+    {
+        Key {
+            name: name.into(),
+            labels: Vec::new(),
+        }
+    }
+
+    /// Creates a `Key` from a name and vector of `Label`s.
+    pub fn from_name_and_labels<N, L>(name: N, labels: L) -> Self
+    where
+        N: Into<ScopedString>,
+        L: IntoLabels,
+    {
+        Key {
+            name: name.into(),
+            labels: labels.into_labels(),
+        }
+    }
+
+    /// Adds a new set of labels to this key.
+    ///
+    /// New labels will be appended to any existing labels.
+    pub fn add_labels(&mut self, new_labels: Vec<Label>) {
+        self.labels.extend(new_labels);
+    }
+
+    /// Name of this key.
+    pub fn name(&self) -> ScopedString {
+        self.name.clone()
+    }
+
+    /// Labels of this key, if they exist.
+    pub fn labels(&self) -> Iter<Label> {
+        self.labels.iter()
+    }
+
+    /// Maps the name of this `Key` to a new name.
+    pub fn map_name<F>(self, f: F) -> Self
+    where
+        F: FnOnce(ScopedString) -> ScopedString,
+    {
+        Key {
+            name: f(self.name),
+            labels: self.labels,
+        }
+    }
+
+    /// Consumes this `Key`, returning the name and any labels.
+    pub fn into_parts(self) -> (ScopedString, Vec<Label>) {
+        (self.name, self.labels)
+    }
+}
+
+impl fmt::Display for Key {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.labels.is_empty() {
+            true => write!(f, "Key({}", self.name),
+            false => {
+                let kv_pairs = self
+                    .labels
+                    .iter()
+                    .map(|label| format!("{} = {}", label.0, label.1))
+                    .collect::<Vec<_>>();
+                write!(f, "Key({}, [{}])", self.name, kv_pairs.join(", "))
+            }
+        }
+    }
+}
+
+impl From<String> for Key {
+    fn from(name: String) -> Key {
+        Key::from_name(name)
+    }
+}
+
+impl From<&'static str> for Key {
+    fn from(name: &'static str) -> Key {
+        Key::from_name(name)
+    }
+}
+
+impl From<ScopedString> for Key {
+    fn from(name: ScopedString) -> Key {
+        Key::from_name(name)
+    }
+}
+
+impl<K, L> From<(K, L)> for Key
+where
+    K: Into<ScopedString>,
+    L: IntoLabels,
+{
+    fn from(parts: (K, L)) -> Key {
+        let labels = parts.1.into_labels();
+        Key::from_name_and_labels(parts.0, labels)
+    }
+}
+
+/// A value that can be converted to `Label`s.
+pub trait IntoLabels {
```

Comment:
Whats the advantage of this trait vs. using something like:
```rust
I: IntoIterator<Item = E>,
E: Into<Label>,
```
Then you could do something like:
```rust
impl From<(K, V)> for Label
where
  K: Into<ScopedString>,
  V: Into<ScopedString>,
````
So that you can convert an iterator of `(K, V)` into `Vec<Label>` like you do later.

Reply:
This is related to the key flexibility we want to provide.  I've been tweaking this stuff a lot so my explanation here could actually be wrong and it might be possible, but:

- we can't implement traits for a foreign type -- `impl Into<...> for T where T: Iterator`, etc -- so we added `IntoLabels` and then derive all the things that can fulfill it, which would be an existing vector of `Label`s or a referenced to an iterator of raw label pairs -- `(K, V) where K/V: Into<ScopedString>`.

This gives us the ability to derive an `Into<Key>` for `(N, L)` where `N` is the metric name and `L` fulfills one of the `IntoLabels` implementations.  Technically, since `Label` is public, we could just focus the constructor methods on handling the generic implementations, but to be honest, I just didn't really think of it.

Without sketching it out, I'm not sure if one is particularly better than the other?

I think with your idea, at least, I could make the `IntoLabels for &'a T` case cleaner by just requiring the item be to `Into<Label>` instead of handling the constraints directly.

Thoughts?

Reply:
I think the main advantage of using the `IntoIterator` approach is that it is more interoperable. For instance, a user that uses a non-std container, lets say `SmallVec<Label>`, would need to convert its data into a `Vec<Label>` first. You also get a lot of stuff for free since almost every std container implements `IntoIterator`, including a `HashMap` which conveniently implements `IntoIterator<Item=(K,V)>`.

Reply:
Yeah, let me try and rework this.  I'm still not sure if I can get rid of `IntoLabels` but it should be clearer once I play around with the code.

Reply:
I think I have a workable solution to this, but it's going to require some macro adjustments that involve more advanced macro syntax, so this is going to take me a little longer than just swapping the generic constraints. :)

Reply:
I spent a lot of time trying to wrangle with the types, based on the common use cases, and here's why I'm deciding to keep `IntoLabels` after all:

The biggest issue came down to the fact that the there's a fundamental tension between callers who use macros (or the new `labels!` macro I added) which can generate a `Vec<Label>` directly and users who end up doing `&[("key", "value")]`, which has to generalize each tuple as `Into<Label>` or `K/V: Into<ScopedString>`.

If we had all of our methods take `IntoIterator<Item = Into<Label>>`, we'd have to unnecessarily allocate/copy a new vector even if we already had `Vec<Label>`, which is a performance penalty I'm not willing to pay.

Right now, the facade macros and the `labels!` helper macro can generate `Vec<Label>` for a user, which should over 99% of the use cases.  For the people that have a custom structure, there's `impl IntoLabels for &T where T: IntoIterator<Item = Into<Label>>` so they simply need to give us a reference to their container, and we'll do the copy for them.

It's not perfect, but I think it covers the intended use cases while still allowing flexibility.

Reply:
If `IntoIterator<Item = Into<Label>>` does allocate like you claim, then this is the right decision. However I'm not sure it actually does. I'll make sure.

Reply:
After looking at the generated assembly for `rustc 1.35`, I can confirm that it allocates. So I agree with your decision.